state-sync/storage-service/server/src/lib.rs

```diff
@@ -166,25 +373,144 @@ impl<T: StorageReaderInterface> StorageServiceServer<T> {
             // avoid starving other async tasks on the same runtime.
             let storage = self.storage.clone();
             let cached_storage_server_summary = self.cached_storage_server_summary.clone();
+            let data_subscriptions = self.data_subscriptions.clone();
             let lru_storage_cache = self.lru_storage_cache.clone();
+            let time_service = self.time_service.clone();
             self.bounded_executor
                 .spawn_blocking(move || {
-                    let response =
-                        Handler::new(storage, cached_storage_server_summary, lru_storage_cache)
-                            .call(protocol, request);
-                    log_storage_response(&response);
-                    response_sender.send(response);
+                    Handler::new(
+                        cached_storage_server_summary,
+                        data_subscriptions,
+                        lru_storage_cache,
+                        storage,
+                        time_service,
+                    )
+                    .process_request_and_respond(
+                        peer,
+                        protocol,
+                        request,
+                        response_sender,
+                    );
                 })
                 .await;
         }
     }
 }
 
+/// Identifies the data subscriptions that can be handled now.
+/// Returns the list of peers that made those subscriptions
+/// alongside the ledger info at the target version for the peer.
+fn get_peers_with_ready_subscriptions<T: StorageReaderInterface>(
+    cached_storage_server_summary: Arc<RwLock<StorageServerSummary>>,
+    data_subscriptions: Arc<Mutex<HashMap<AccountAddress, DataSubscriptionRequest>>>,
+    storage: T,
+) -> Result<Vec<(AccountAddress, LedgerInfoWithSignatures)>, Error> {
+    // Fetch the latest storage summary and highest synced version
+    let latest_storage_summary = cached_storage_server_summary.read().clone();
+    let highest_synced_ledger_info = match latest_storage_summary.data_summary.synced_ledger_info {
```

Comment:
@JoshLind - May be I am missing something here but on a high level, why do the client need to inform the server of the `highest_synced_ledger_info` - in the subscription model, the server can figure that out based on last update server sent to the client?

Reply:
My point was around why this is not a "pure subscription" model, which you already answered below.